candid/src/de.rs

```diff
@@ -777,3 +777,57 @@ impl<'de, 'a> de::VariantAccess<'de> for Compound<'a, 'de> {
         }
     }
 }
+
+pub trait DecodeArguments<'a>: Sized {
+    fn decode_arguments(_de: IDLDeserialize<'a>) -> Result<(IDLDeserialize<'a>, Self)>;
+}
+
+// // Is this a sensible impl?
+impl<'a> DecodeArguments<'a> for () {
+    fn decode_arguments(de: IDLDeserialize<'a>) -> Result<(IDLDeserialize<'a>, ())> {
+        Ok((de, ()))
+    }
+}
+
+// This is all pretty mechanical, perhaps we could use code gen?
+impl<'a, A1: Deserialize<'a>> DecodeArguments<'a> for (A1,) {
```

Comment:
It's a bit un-natural to always write one argument as `(v,)`. Do you have better ideas, e.g. impl the trait for `T: Deserialize`?

Reply:
Having that impl leads to ambiguities since `(A,B)` could be a single argument of a tuple or two arguments of type A and B.
An option is just to have a function like `decode_arg` which does something like
```rust
let (a,) = decode_args(inp);
a
```

Reply:
And yeah, I'll switch it over to codegen if it isn't too hairy